### PR TITLE
Implement VS Code tree view checkbox API.

### LINKED ISSUE #12695: [vscode] Support Tree checkbox API
According to vscode documentation, API about Tree Checkbox API has been finalized.

https://code.visualstudio.com/updates/v1_80#_tree-checkbox-api

Results of the report:
TreeCheckboxAPI

### PR DESCRIPTION
**Pull request by tsmaeder**
#### What it does
Implements the ability to have check boxes on tree views in plugins

Fixes #12695

Contributed on behalf of STMicroelectronics

#### How to test
Install the attached VS Code extension. It contributes a a view called "Checkbox API Test". Open the view if necessary. Play around with checking/unchecking the checkboxes. The system automatically checks/unchecks parent and child checkboxes: see the VS Code doc linked in the issue for details.
The extension also contributes a preference "Manage Checkbox State Manually". When this preference is on, the automatic checking/unchecking is disabled. The extensions writes to the log, when it is notified of checkbox state changes. There is also a context menu action to check/uncheck the tree items programmatically. 

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by tsmaeder** [comment]
Here's the vsix:

[checkboxapi-0.0.1.zip](https://github.com/eclipse-theia/theia/files/12357455/checkboxapi-0.0.1.zip)

and the source lives here: 

https://github.com/tsmaeder/checkboxapi/

File: packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx [hunk 1]
```diff
@@ -272,6 +273,40 @@ export class PluginTree extends TreeImpl {
...
+        function findParentsToChange(child: TreeNode, nodes: TreeNode[]): void {
+            if (child.parent) {
+                if (child.parent.checkboxInfo !== undefined && child.parent.checkboxInfo.checked !== checked) {
+                    if (!checked || child.parent.children.every(candidate => candidate.checkboxInfo?.checked || candidate.checkboxInfo === undefined || candidate === child)) {
+                        nodes.push(child.parent);
+                        findParentsToChange(child.parent, nodes);
+                    }
+                }
+            }
+        }
```
Comment (msujew) [inline packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx:286]: Minor: Quite a lot of nesting in here:
```suggestion
            if (child.parent?.checkboxInfo !== undefined 
             && child.parent.checkboxInfo.checked !== checked
             && (!checked || child.parent.children.every(candidate => 
                 candidate.checkboxInfo?.checked 
                 || candidate.checkboxInfo === undefined 
                 || candidate === child
             ))) {
                 nodes.push(child.parent);
                 findParentsToChange(child.parent, nodes);
            }
        }
```

File: packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx [hunk 2]
```diff
@@ -272,6 +273,40 @@ export class PluginTree extends TreeImpl {
...
+
+        function findChildrenToChange(parent: TreeNode, nodes: TreeNode[]): void {
+            if (CompositeTreeNode.is(parent)) {
+                parent.children.forEach(child => {
+                    if (child.checkboxInfo !== undefined && child.checkboxInfo.checked !== checked) {
```
Comment (msujew) [inline packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx:291]: Minor:
```suggestion
                    if (child.checkboxInfo?.checked !== checked) {
```

**Review by msujew — APPROVED** [review]
Looking quite good 👍 

* The API is marked as supported in the comparator report
* The checkbox rendering works as expected
* Changing the preference sets the checkboxes into manual mode (requires reload of the plugin after changing the preference)

I have two minor comments below. Should be good to go afterwards 👍

File: packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx (hunk 2 above)
Comment (msujew) [reply packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx:288]: Right, nevermind. I missunderstood the reason why the `child.checkboxInfo !== undefined` exists in the first place. Nothing to see here... :)

**Comment by tsmaeder** [comment]
Thanks for the review.

**Pull request merged**